### PR TITLE
feat(uat): added to t103 scenario userProperties and related steps

### PR DESCRIPTION
**Pull request by auarbekov**
**Issue #, if available:**
https://klika-tech.atlassian.net/browse/GGMQ-186
feat(uat): added to t103 scenario userProperties and related steps

**Description of changes:**
- Added t104 scenario

**Why is this change necessary:**
In GGMQ-1-T104 we testing ability to user user properties for MQTT 5.0

**How was this change tested:**
Scenarios run manually and on CI

**Test results:**
For all clients:
```
Given my device is registered as a Thing....................................passed
And my device is running Greengrass.........................................passed
When I create a Greengrass deployment with components.......................passed
And I create client device "publisher"......................................passed
And I create client device "subscriber".....................................passed
When I associate "subscriber" with ggc......................................passed
When I associate "publisher" with ggc.......................................passed
And I update my Greengrass deployment configuration, setting the component aws.greengrass.clientdevices.Auth configuration to:.passed
And I update my Greengrass deployment configuration, setting the component aws.greengrass.client.Mqtt5JavaSdkClient configuration to:.passed
And I deploy the Greengrass deployment configuration........................passed
Then the Greengrass deployment is COMPLETED on the device after 5 minutes...passed
And the aws.greengrass.clientdevices.mqtt.EMQX log on the device contains the line "is running now!." within 1 minutes.passed
Then I discover core device broker as "localMqttBroker1" from "publisher" in OTF.passed
Then I discover core device broker as "localMqttBroker2" from "subscriber" in OTF.passed
And I connect device "publisher" on aws.greengrass.client.Mqtt5JavaSdkClient to "localMqttBroker1" using mqtt "v5".passed
And I connect device "subscriber" on aws.greengrass.client.Mqtt5JavaSdkClient to "localMqttBroker2" using mqtt "v5".passed
And I set MQTT subscribe 'retain as published' flag to false................passed
When I subscribe "subscriber" to "iot_data_0" with qos 0....................passed
And I set MQTT publish 'retain' flag to false...............................passed
And I set the 'retain' flag in expected received messages to false..........passed
When I publish from "publisher" to "iot_data_0" with qos 0 and message "Hello world1".passed
And message "Hello world1" received on "subscriber" from "iot_data_0" topic within 5 seconds.passed
And I set MQTT publish 'retain' flag to true................................passed
When I publish from "publisher" to "iot_data_0" with qos 0 and message "Hello world2".passed
And message "Hello world2" received on "subscriber" from "iot_data_0" topic within 5 seconds.passed
And I set MQTT subscribe 'retain as published' flag to true.................passed
When I subscribe "subscriber" to "iot_data_1" with qos 0....................passed
And I set MQTT publish 'retain' flag to false...............................passed
And I set the 'retain' flag in expected received messages to false..........passed
When I publish from "publisher" to "iot_data_1" with qos 0 and message "Hello world3".passed
And message "Hello world3" received on "subscriber" from "iot_data_1" topic within 5 seconds.passed
And I set MQTT publish 'retain' flag to true................................passed
And I set the 'retain' flag in expected received messages to true...........passed
When I publish from "publisher" to "iot_data_1" with qos 0 and message "Hello world4".passed
And message "Hello world4" received on "subscriber" from "iot_data_1" topic within 5 seconds.passed
And I add MQTT 'user property' with key "type" and value "json" to transmit.passed
And I add MQTT 'user property' with key "region" and value "Asia" to transmit.passed
And I add MQTT 'user property' with key "type" and value "json" to receive..passed
And I add MQTT 'user property' with key "region" and value "Asia" to receive.passed
When I subscribe "subscriber" to "iot_data_2" with qos 0....................passed
When I publish from "publisher" to "iot_data_2" with qos 0 and message "Expected userProperties are received".passed
And message "Expected userProperties are received" received on "subscriber" from "iot_data_2" topic within 5 seconds.passed
And I clear message storage.................................................passed
And I clear MQTT 'user properties' to transmit..............................passed
And I clear MQTT 'user properties' to receive...............................passed
And I add MQTT 'user property' with key "agent-control" and value "control" to receive.passed
And I add MQTT 'user property' with key "timezone" and value "GMT6" to receive.passed
When I subscribe "subscriber" to "iot_data_2" with qos 0....................passed
When I publish from "publisher" to "iot_data_2" with qos 0 and message "Expected userProperties are not received".passed
And message "Expected userProperties are not received" is not received on "subscriber" from "iot_data_2" topic within 5 seconds.passed
And I clear message storage.................................................passed
And I clear MQTT 'user properties' to transmit..............................passed
And I clear MQTT 'user properties' to receive...............................passed
And I add MQTT 'user property' with key "agent-control" and value "control" to transmit.passed
And I add MQTT 'user property' with key "timezone" and value "GMT6" to transmit.passed
When I subscribe "subscriber" to "iot_data_3" with qos 0....................passed
When I publish from "publisher" to "iot_data_3" with qos 0 and message "Ignore userProperties".passed
And message "Ignore userProperties" received on "subscriber" from "iot_data_3" topic within 5 seconds.passed
And I clear message storage.................................................passed
And I clear MQTT 'user properties' to transmit..............................passed
And I clear MQTT 'user properties' to receive...............................passed
When I subscribe "subscriber" to "iot_data_4" with qos 0....................passed
When I publish from "publisher" to "iot_data_4" with qos 0 and message "Without userProperties".passed
And message "Without userProperties" received on "subscriber" from "iot_data_4" topic within 5 seconds.passed
And I disconnect device "subscriber" with reason code 0.....................passed
And I disconnect device "publisher" with reason code 0......................passed
]]>
```

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by eschastlivtsev — DISMISSED** [review]

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 1]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
         log.info("Storage was cleared");
     }
 
+    /**
+     * Sets MQTT user Properties to transmit.
+     *
+     * @param key the key of userProperties property.
+     * @param value the value of userProperties property.
+     */
+    @And("I set MQTT user property with key {string} and value {string} to transmit")
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: Take user property in single quotes.

Better 
"I add MQTT 'user property' with key and value to transmit"

Rename method to addTxUserProperty

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 2]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+                .setValue(value)
+                .build();
+        txUserProperties.add(properties);
+    }
+
+    /**
+     * Clear MQTT user properties to transmit.
+     *
+     */
+    @And("I clear MQTT user properties to transmit")
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: Take 'user properties' to single quotes

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 3]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+        txUserProperties = null;
+    }
+
+    /**
+     * Sets MQTT user Properties to receive.
+     *
+     * @param key the key of userProperties property.
+     * @param value the value of userProperties property.
+     */
+    @And("I set MQTT user property with key {string} and value {string} to receive")
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: add and quotes

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 4]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+                .setValue(value)
+                .build();
+        rxUserProperties.add(properties);
+    }
+
+    /**
+     * Clear MQTT user properties to receive.
+     *
+     */
+    @And("I clear MQTT user properties to receive")
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: Quotes

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 5]
```diff
@@ -1204,3 +1204,99 @@ Feature: GGMQ-1
...
+    Then the Greengrass deployment is COMPLETED on the device after 5 minutes
+    And the aws.greengrass.clientdevices.mqtt.EMQX log on the device contains the line "is running now!." within 1 minutes
+
+    Then I discover core device broker as "localMqttBroker1" from "publisher" in OTF
+    Then I discover core device broker as "localMqttBroker2" from "subscriber" in OTF
+    And I connect device "publisher" on <agent> to "localMqttBroker1" using mqtt "<mqtt-v>"
+    And I connect device "subscriber" on <agent> to "localMqttBroker2" using mqtt "<mqtt-v>"
+
+    And I set MQTT user property with key "type" and value "json" to transmit
+    And I set MQTT user property with key "type" and value "json" to receive
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature]: Better add at least two properties to test code.

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 6]
```diff
@@ -1204,3 +1204,99 @@ Feature: GGMQ-1
...
+    When I publish from "publisher" to "iot_data_1" with qos 0 and message "Hello world"
+    And message "Hello world" received on "subscriber" from "iot_data_1" topic within 5 seconds
+
+    And I clear message storage
+    And I clear MQTT user properties to transmit
+    And I clear MQTT user properties to receive
+    And I set MQTT user property with key "agent-control" and value "control" to transmit
+    When I subscribe "subscriber" to "iot_data_2" with qos 0
+    When I publish from "publisher" to "iot_data_2" with qos 0 and message "Hello world1"
+    And message "Hello world1" is not received on "subscriber" from "iot_data_2" topic within 5 seconds
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature]: Why is not received ?
When `received` properties is null that means no filter for properties is applied.

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 7]
```diff
@@ -143,7 +143,10 @@ public class MqttControlSteps {
 
 
     /** Actual list of user properties to transmit. */
-    private static final List<Mqtt5Properties> txUserProperties = null;
+    private static List<Mqtt5Properties> txUserProperties = null;
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java:146]: 1. Where txUserProperties is used ?
2. Create DEFAULT_TX_USER_PROPERTIES equals to null

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 8]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+     *
+     * @param key the key of userProperties property.
+     * @param value the value of userProperties property.
+     */
+    @And("I set MQTT user property with key {string} and value {string} to transmit")
+    public void setTxUserProperties(String key, String value) {
+        if (txUserProperties == null) {
+            txUserProperties = new ArrayList<>();
+        }
+        Mqtt5Properties properties = Mqtt5Properties.newBuilder()
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: rename to property (here is single property)

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 9]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+    }
+
+    /**
+     * Sets MQTT user Properties to receive.
+     *
+     * @param key the key of userProperties property.
+     * @param value the value of userProperties property.
+     */
+    @And("I set MQTT user property with key {string} and value {string} to receive")
+    public void setRxUserProperties(String key, String value) {
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: rename to addRxUserProperty()

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java [hunk 10]
```diff
@@ -674,6 +678,63 @@ public void clearStorage() {
...
+        if (rxUserProperties == null) {
+            rxUserProperties = new ArrayList<>();
+        }
+        Mqtt5Properties properties = Mqtt5Properties.newBuilder()
```
Comment (bgklika) [inline uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java]: rename to property

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 11]
```diff
@@ -1204,3 +1204,99 @@ Feature: GGMQ-1
     Examples:
       | mqtt-v | name        | agent                                     | recipe                  |
       | v5     | paho-java   | aws.greengrass.client.Mqtt5JavaPahoClient | client_java_paho.yaml   |
+
+
+  @GGMQ-1-T104
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature]: Please join with T103. 
Michael worred about time spend to run scenarios.

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 12]
```diff
@@ -1204,3 +1204,99 @@ Feature: GGMQ-1
...
+
+    @mqtt5 @mosquitto-c
+    Examples:
+      | mqtt-v | name        | agent                                     | recipe                  |
+      | v5     | mosquitto-c | aws.greengrass.client.MqttMosquittoClient | client_mosquitto_c.yaml |
+
+    @mqtt5 @paho-java
+    Examples:
+      | mqtt-v | name        | agent                                     | recipe                  |
+      | v5     | paho-java   | aws.greengrass.client.Mqtt5JavaPahoClient | client_java_paho.yaml   |
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature]: \n at end of line is missing

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/addon/EventFilter.java [hunk 13]
```diff
@@ -197,6 +202,17 @@ public Builder withRetain(Boolean retain) {
             return this;
         }
 
+        /**
+         * Sets user properties.
```
Comment (bgklika) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/addon/EventFilter.java:206]: Sets or resets

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/addon/MqttMessageEvent.java [hunk 14]
```diff
@@ -118,6 +125,17 @@ private boolean isRetainMatched(Boolean retain) {
         return retain == null || retain == message.getRetain();
     }
 
+    private boolean isUserPropertiesMatched(List<Mqtt5Properties> userProperties) {
+        if (userProperties == null) {
```
Comment (bgklika) [inline uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/implementation/addon/MqttMessageEvent.java]: If argument userProperties (from filter) is null that means - we does not want to check user properties in received message.

Please add unit tests to check it.

Unit tests for retain is already added, another example is here:
https://github.com/Klika-Tech/aws-greengrass-client-device-auth/pull/84/files#diff-adbf31335af75d632627d69f28d8d169ce063414224b5916799a422f6f9c7e0cR359

**Review by bgklika — CHANGES_REQUESTED** [review]

File: uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/addon/EventFilter.java (hunk 13 above)
Comment (auarbekov) [reply uat/mqtt-client-control/src/main/java/com/aws/greengrass/testing/mqtt/client/control/api/addon/EventFilter.java:206]: It's builder's method, that's why I wrote sets.

File: uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java (hunk 7 above)
Comment (bgklika) [reply uat/testing-features/src/main/java/com/aws/greengrass/steps/MqttControlSteps.java:146]: txUserProperties is still not used to connect, subscribe, publish, unsubscribe, disconnect requests.

**Review by bgklika — CHANGES_REQUESTED** [review]

**Review by bgklika — APPROVED** [review]

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 15]
```diff
@@ -1187,6 +1187,50 @@ Feature: GGMQ-1
...
+    And I clear message storage
+    And I clear MQTT 'user properties' to transmit
+    And I clear MQTT 'user properties' to receive
+    And I add MQTT 'user property' with key "agent-control" and value "control" to receive
+    And I add MQTT 'user property' with key "timezone" and value "GMT6" to receive
+    When I subscribe "subscriber" to "iot_data_2" with qos 0
+    When I publish from "publisher" to "iot_data_2" with qos 0 and message "Expected userProperties are not received"
+    And message "Expected userProperties are not received" is not received on "subscriber" from "iot_data_2" topic within 5 seconds
+
+    # 5. test case when publish 'user properties' are empty.
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature:1213]: The same as 4. by description

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 16]
```diff
@@ -1187,6 +1187,50 @@ Feature: GGMQ-1
...
+    And I clear message storage
+    And I clear MQTT 'user properties' to transmit
+    And I clear MQTT 'user properties' to receive
+    And I add MQTT 'user property' with key "agent-control" and value "control" to transmit
+    And I add MQTT 'user property' with key "timezone" and value "GMT6" to transmit
+    When I subscribe "subscriber" to "iot_data_3" with qos 0
+    When I publish from "publisher" to "iot_data_3" with qos 0 and message "Ignore userProperties"
+    And message "Ignore userProperties" received on "subscriber" from "iot_data_3" topic within 5 seconds
+
+    # 6. test case when publish 'user properties' are empty.
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature:1224]: The same description as 4.

File: uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature [hunk 17]
```diff
@@ -1187,6 +1187,50 @@ Feature: GGMQ-1
...
+    #  In that case 'user properties' on receive should be equal to 'user properties' value on publish.
+
+    And I clear message storage
+    And I clear MQTT 'user properties' to transmit
+    And I clear MQTT 'user properties' to receive
```
Comment (bgklika) [inline uat/testing-features/src/main/resources/greengrass/features/ggmq-1.feature:1229]: It is default case and tested in all other scenarios

**Pull request closed without merging**